### PR TITLE
basic framework to run integration tests

### PR DESCRIPTION
**Pull request by individual-it**
basic framework and first small set of integration tests
`make test-php-integration` should make it run locally

1. ocis is started with keycloak and some well known users
2. tokens are fetched using oidc flow type `Resource Owner Password Flow` setting `grant_type=password`
3. tokens are used to do various operations through the SDK

### CODE REVIEW

File: tests/integration/Owncloud/OcisSdkPhp/OcisTest.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    private const UUID_REGEX_PATTERN = '/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}\$[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i';
+    private string $ocisUrl = 'https://ocis.sdk.test:9009';
+    private ?string $tokenUrl = null;
+    private ?Client $guzzleClient = null;
+    /**
+     * @var array <string>
+     */
+    private $createdDrives = [];
+
+    public function setUp(): void
```
Comment (amrita-shrestha) [inline tests/integration/Owncloud/OcisSdkPhp/OcisTest.php:23]: why you name this setUp?

Comment (individual-it) [reply tests/integration/Owncloud/OcisSdkPhp/OcisTest.php:23]: This is how the method, that runs before every test, has to  be called in PHPUnit
https://docs.phpunit.de/en/10.4/fixtures.html

File: src/Drive.php [hunk 2]
```diff
@@ -143,15 +155,61 @@ public function getRawData(): mixed
         return $this->apiDrive->jsonSerialize();
     }
 
+    /**
+     * Deletes the current drive irreversibly.
+     * A drive can only be deleted if its has been disabled before.
```
Comment (phil-davis) [inline src/Drive.php:160]:
```suggestion
     * A drive can only be deleted if it has already been disabled.
```

File: src/Drive.php [hunk 3]
```diff
@@ -143,15 +155,61 @@ public function getRawData(): mixed
...
+            $apiInstance->deleteDrive($this->getId());
+        } catch (ApiException $e) {
+            throw ExceptionHelper::getHttpErrorException($e);
+        }
     }
 
+    /**
+     * Disables the current drive without deleting it.
+     * Disabling a drive is the prerequisite for deleting it.
+     * Calling this function on a drive that is not disabled will have no effect.
```
Comment (phil-davis) [inline src/Drive.php:191]:
```suggestion
     * Calling this function on a drive that is already disabled will have no effect.
```

**Review by phil-davis — COMMENTED** [review]
Just a couple of comments

**Comment by phil-davis** [comment]
```
$ make test-php-integration
cd tests/integration && docker compose up -d --wait
[+] Running 4/4
 ✔ Network integration_default       Created                                                                          0.1s 
 ✔ Container integration-postgres-1  Healthy                                                                          0.1s 
 ✔ Container integration-ocis-1      Started                                                                          0.1s 
 ✔ Container integration-keycloak-1  Healthy                                                                          0.1s 
container integration-ocis-1 is unhealthy
make: *** [Makefile:54: run-ocis-with-keycloak] Error 1
```
Any clues about what I should do to get a clean start to run this?

**Review by phil-davis — APPROVED** [review]
LGTM - I "just" have some problem running it locally on my laptop.
But I assume that @individual-it can run the integration tests, and I need to find out what is the local problem on my system.

**Comment by individual-it** [comment]
@phil-davis you might have ran into https://github.com/owncloud/ocis/issues/7438 
try to run:
```
cd tests/integration
docker compose down -v --remove-orphans --rmi all
sudo systemctl restart docker
```

that fixed it for me for now

**Comment by phil-davis** [comment]
```
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk$ cd tests/integration/
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk/tests/integration$ docker compose down -v --remove-orphans --rmi all
[+] Running 8/8
 ✔ Container integration-ocis-1       Removed                                                                         0.5s 
 ✔ Container integration-keycloak-1   Removed                                                                         0.6s 
 ✔ Container integration-postgres-1   Removed                                                                         0.3s 
 ✔ Volume integration_keycloak_data   Removed                                                                         0.0s 
 ✔ Image owncloud/ocis:latest         Removed                                                                         0.1s 
 ✔ Image integration-keycloak:latest  Removed                                                                         0.3s 
 ✔ Image postgres:alpine              Removed                                                                         0.3s 
 ✔ Network integration_default        Removed                                                                         0.1s 
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk/tests/integration$ sudo systemctl restart docker
[sudo] password for phil: 
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk/tests/integration$ cd ..
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk/tests$ cd ..
phil@phil-Inspiron-5468:~/git/owncloud/ocis-php-sdk$ make test-php-integration 
cd tests/integration && docker compose up -d --wait
[+] Running 14/14
 ✔ postgres 8 layers [⣿⣿⣿⣿⣿⣿⣿⣿]      0B/0B      Pulled                                                               46.3s 
   ✔ 96526aa774ef Pull complete                                                                                       1.9s 
   ✔ 1c7e9d0e60f3 Pull complete                                                                                       0.9s 
   ✔ 9538d5302eee Pull complete                                                                                       0.9s 
   ✔ ddefdf18815a Pull complete                                                                                      39.7s 
   ✔ 483d47e7b846 Pull complete                                                                                       2.9s 
   ✔ 4f69e6f6f92e Pull complete                                                                                       2.8s 
   ✔ 6ad5f543b2cb Pull complete                                                                                       3.7s 
   ✔ 3f41102d5734 Pull complete                                                                                       3.7s 
 ✔ ocis 4 layers [⣿⣿⣿⣿]      0B/0B      Pulled                                                                       27.9s 
   ✔ a2e76cdd5b16 Pull complete                                                                                       5.5s 
   ✔ df1d9526990f Pull complete                                                                                       4.8s 
   ✔ b9ff7f31b7f0 Pull complete                                                                                       6.3s 
   ✔ 5cb9e3c39fd6 Pull complete                                                                                      22.6s 
[+] Building 80.6s (10/10) FINISHED                                                                         docker:default
 => [keycloak internal] load build definition from Dockerfile                                                         0.0s
 => => transferring dockerfile: 877B                                                                                  0.0s
 => [keycloak internal] load .dockerignore                                                                            0.0s
 => => transferring context: 2B                                                                                       0.0s
 => [keycloak internal] load metadata for quay.io/keycloak/keycloak:22.0.4                                            2.6s
 => [keycloak internal] load metadata for docker.io/library/ubuntu:22.04                                              2.1s
 => [keycloak gen-certs 1/2] FROM docker.io/library/ubuntu:22.04@sha256:9b8dec3bf938bc80fbe758d856e96fdfab5f56c39d44  0.0s
 => [keycloak stage-1 1/3] FROM quay.io/keycloak/keycloak:22.0.4@sha256:20c2bd474f20a3f89b7c652b9c4e5a54701301beddc  77.3s
 => => resolve quay.io/keycloak/keycloak:22.0.4@sha256:20c2bd474f20a3f89b7c652b9c4e5a54701301beddcee7deb00ec88d73cbb  0.0s
 => => sha256:fc65979ef8be73a3fb8c59ee66f26d261537963bc0670dbff0d7134489e519a2 164.66MB / 164.66MB                   75.2s
 => => sha256:20c2bd474f20a3f89b7c652b9c4e5a54701301beddcee7deb00ec88d73cbbab5 1.61kB / 1.61kB                        0.0s
 => => sha256:12a2250bf731038caa5cda9f64b316af223f27de2afeef20974eb777601bc6f2 1.06kB / 1.06kB                        0.0s
 => => sha256:8ef105ff6a8d4a6f58978e20e0cdf14fb1dadbe830ba8a68123ac575d8d8e339 6.06kB / 6.06kB                        0.0s
 => => sha256:9e167103aefa56af4c39359687f54fdd263ca85bc946e770d92c054df7ada57c 7.69MB / 7.69MB                        8.1s
 => => sha256:139b6b161a2f250f2d681753028510d391da42fd60ce3404582f53795e4a17af 76.71MB / 76.71MB                     49.3s
 => => extracting sha256:9e167103aefa56af4c39359687f54fdd263ca85bc946e770d92c054df7ada57c                             0.5s
 => => sha256:160dfdb51fefe875e1bff71f46e04f457f235df333e195212c2f70e029ee7822 550B / 550B                            9.0s
 => => extracting sha256:139b6b161a2f250f2d681753028510d391da42fd60ce3404582f53795e4a17af                             1.6s
 => => extracting sha256:fc65979ef8be73a3fb8c59ee66f26d261537963bc0670dbff0d7134489e519a2                             1.4s
 => => extracting sha256:160dfdb51fefe875e1bff71f46e04f457f235df333e195212c2f70e029ee7822                             0.0s
 => CACHED [keycloak gen-certs 2/2] RUN apt update && apt install -y openssl &&     mkdir -p /certs && cd /certs &&   0.0s
 => [keycloak stage-1 2/3] COPY --from=gen-certs /certs /opt/keycloak/certs                                           0.2s
 => [keycloak stage-1 3/3] RUN chown -R keycloak:root /opt/keycloak/certs                                             0.4s
 => [keycloak] exporting to image                                                                                     0.1s
 => => exporting layers                                                                                               0.1s
 => => writing image sha256:3a03c8eb57a27509bf93e9743b474b68a8eeac603f281b4a5fee5e5fa3a50400                          0.0s
 => => naming to docker.io/library/integration-keycloak                                                               0.0s
[+] Running 5/5
 ✔ Network integration_default         Created                                                                        0.1s 
 ✔ Volume "integration_keycloak_data"  Created                                                                        0.0s 
 ✔ Container integration-postgres-1    Healthy                                                                        0.1s 
 ✔ Container integration-ocis-1        Started                                                                        0.1s 
 ✔ Container integration-keycloak-1    Healthy                                                                        0.1s 
container integration-ocis-1 is unhealthy
make: *** [Makefile:54: run-ocis-with-keycloak] Error 1
```

It seems that there is some other problem. I will leave it for some other day.

**Comment by individual-it** [comment]
@phil-davis  After a lot of debugging this could also be a firewall issue. To confirm try to disable temporary the firewall (e.g. ufw) on your system

**Pull request merged**